src/main/java/edu/ucla/library/iiif/fester/handlers/PostThumbnailsHandler.java

```diff
@@ -0,0 +1,292 @@
+        final HttpRequest<JsonObject> request;
+        request = WebClient.create(myVertx)
+          .getAbs(manifestURL)
+          .putHeader("Accept", Constants.JSON_MEDIA_TYPE)
+          .as(BodyCodec.jsonObject())
+          .expect(ResponsePredicate.SC_OK);
+        if (manifestURL.startsWith("https")) {
+            request.ssl(true);
+        }
+        request.send(asyncResult -> {
+            if (asyncResult.succeeded()) {
+                final JsonObject manifestBody = asyncResult.result().body();
+                final String context = manifestBody.getString(CONTEXT);
+                final int thumbIndex = ThumbnailUtils.findThumbHeaderIndex(aCsvList.get(0));
+                if (context.contains(Constants.CONTEXT_V2)) {
+                    addV2Thumb(thumbIndex, aRowIndex, manifestBody, aCsvList);
+                } else if (context.contains(Constants.CONTEXT_V3)) {
+                    addV3Thumb(thumbIndex, aRowIndex, manifestBody, aCsvList);
+                } else {
+                    LOGGER.info(LOGGER.getMessage(MessageCodes.MFS_167, manifestURL));
+                }
+                promise.complete();
+            } else {
+                promise.fail(asyncResult.result().statusMessage());
+            }
+        });
+
+        return promise.future();
+
+    }
+
+    /**
+     * Retireve the base thumbnail URL from IIIF V2 presentation manifests.
+     *
+     * @param aColumnIndex The column in the CSV where thumbnail will be added
+     * @param aRowIndex The CSV row being updated
+     * @param aManifest A IIIF work manifest
+     * @param aCsvList A CSV file parsed as a list of string arrays
+     */
+    private void addV2Thumb(final int aColumnIndex, final int aRowIndex,
+                            final JsonObject aManifest, final List<String[]> aCsvList) {
+        final JsonArray canvases = aManifest.getJsonArray(SEQUENCES).getJsonObject(0).getJsonArray(CANVASES);
+        final int canvasIndex = chooseThumbIndex(canvases.size());
+        final String thumbURL = canvases.getJsonObject(canvasIndex).getJsonArray(IMAGE_CONTENT)
+                                .getJsonObject(0).getJsonObject(RESOURCE).getJsonObject(SERVICE)
+                                .getString(Constants.ID_V2);
+        ThumbnailUtils.addThumbnailURL(aColumnIndex, aRowIndex, thumbURL, aCsvList);
+    }
+
+    /**
+     * Retireve the base thumbnail URL from IIIF V3 presentation manifests.
+     *
+     * @param aColumnIndex The column in the CSV where thumbnail will be added
+     * @param aRowIndex The CSV row being updated
+     * @param aManifest A IIIF work manifest
+     * @param aCsvList A CSV file parsed as a list of string arrays
+     */
+    private void addV3Thumb(final int aColumnIndex, final int aRowIndex,
+                            final JsonObject aManifest, final List<String[]> aCsvList) {
+
+        final JsonArray canvases = aManifest.getJsonArray(ITEMS).getJsonObject(0).getJsonArray(ITEMS);
```

Comment:
For v3, the list of canvases is at `manifest["items"]`, not `manifest["items"][0]["items"]`.

Our manifests all (at the moment) have a 1:1 correspondence between canvases and images, so once you choose a  canvas with `canvasIndex` as you're already doing, you can get the image with `canvas["items"][0]["items"][0]["body"]`, then the IIIF service URL with `image["service"][0]["id"]`.

All these layers make a lot more sense in a linked data context (which were only using at a very surface level).

Reply:
This is a V3 work manifest generated by TEST Fester: https://test-iiif.library.ucla.edu/ark%3A%2F21198%2Fzz001d2sw4/manifest; unless I've gone bonkers, it most definitely has `manifest["items"][0]["items"]`.

Reply:
It does, however, the value at `manifest["items"][0]["items"]` is a list of `AnnotationPage`s, not a list of `Canvas`es. For our items, at the moment there is only one AnnotationPage per canvas, but (potentially) multiple Canvases per manifest.

Reply:
Modified logic